### PR TITLE
Refresh format and Kubernetes standards

### PR DESCRIPTION
**Pull request by johnwatson484**
https://eaflood.atlassian.net/browse/PSD-966

### CODE REVIEW

File: README.md
Comment (st3v3nhunt) [inline README.md]: Path needs to be `guides/jlm.md`

File: README.md
Comment (st3v3nhunt) [inline README.md]: Testing is currently under `standards` rather than `guides`.

File: guides/kubernetes/README.md
Comment (st3v3nhunt) [inline guides/kubernetes/README.md]: Path should be just `pod-identity.md`

File: guides/collaboration.md
Comment (st3v3nhunt) [inline guides/collaboration.md:11]: I've not used Google Drive since I started. Am I missing out or could this option be removed altogether?

File: guides/collaboration.md
Comment (st3v3nhunt) [inline guides/collaboration.md:46]: Some of the channels listed are private and will require an invitation, it might be worth making a note this will need to done by an existing member of the channel.

File: guides/developer-laptop-setup/install-vs-code.md
Comment (st3v3nhunt) [inline guides/developer-laptop-setup/install-vs-code.md:20]: nit: `.Net` has been altered to `.NET` in other places.

File: guides/developer-laptop-setup/install-vs-code.md
Comment (st3v3nhunt) [inline guides/developer-laptop-setup/install-vs-code.md:48]: Ending `)` is missing from link.

File: guides/jlm.md
Comment (st3v3nhunt) [inline guides/jlm.md:5]: Typo: `adminstration` ➡️ `administration`

File: guides/jlm.md
Comment (st3v3nhunt) [inline guides/jlm.md:7]: Typo: `accesss` ➡️ `access`

File: guides/shared-assets.md
Comment (st3v3nhunt) [inline guides/shared-assets.md:55]: Typo: `repostory` ➡️ `repository`
Typo: `defininitions` ➡️ `definitions`

File: standards/kubernetes/helm-charts.md
Comment (st3v3nhunt) [inline standards/kubernetes/helm-charts.md:2]: Typo: `Kubenetes` ➡️ `Kubernetes`

File: standards/quick-reference.md
Comment (st3v3nhunt) [inline standards/quick-reference.md]: I thought secrets were loaded into pods via `Secret` Kubernetes resources?
If the bullet read `do not store _credentials_ in a pod...` that would make more sense to me. Is that more accurate? I might be missing something though.

File: standards/source-code.md
Comment (st3v3nhunt) [inline standards/source-code.md:57]: Is it worth having this as a file that can be copied into the repo?

File: standards/source-code.md
Comment (st3v3nhunt) [inline standards/source-code.md]: Typo: `view` ➡️ `review`, at least I think it should be `review`!?

File: standards/version-control.md
Comment (st3v3nhunt) [inline standards/version-control.md:17]: Not really anything to do with this PR but it has reminded me. I think we should extend the version increment to include `package-lock.json` so they are kept in sync. Atm, we have a drift between files, which isn't a problem but when packages are updated the `package-lock.json` gets updated with what the current version is and depending when that happens it might be the very latest version or it might end up being the previous version. If we included the file in the checks, it would force the update.
Just an idea...

File: guides/developer-laptop-setup/README.md
Comment (st3v3nhunt) [inline guides/developer-laptop-setup/README.md]: `AWS` ➡️ `Azure`

**Review by st3v3nhunt — CHANGES_REQUESTED** [review]
A lot of good content in here 👏. I like the definition of guides and standards. I think there is scope to have a level above a standard, something like a mandate or mandated standard i.e. if you want stuff to work it needs to meet the mandated standard. Some examples would be a branch name is a standard but not mandated, nothing isn't going to work that doesn't adhere to it. At least atm, that is how branches work, ofc, in the future it might be the case that only branches meeting that standard will be built and that would be the time when it becomes mandates. Another example would be the naming of the repos, that is really a mandated standard as things will not work if it isn't followed. Something to discuss outside of GitHub issues most likely.

I've added comments about typos and broken links where I've spotted them. There is also some rogue whitespace I noticed when I looked at the changes locally but I've not commented as it isn't visible in the GitHub diff.

A slight aside, I wonder if there should be a style guide for markdown content as part of this repo. My preference for content is the same as code i.e. a max of 80 character wide lines. The majority of content I see doesn't appear to have any limitation. Obviously there are pros and cons to both so whatever is decided I'm OK with, it is more about having a standard. This goes for things like blank lines between headings, double spaces for line breaks, capitalisation of DEFRA, bullets not ending with a period, etc.
What do you think? Something else for wider discussion perhaps?

File: guides/collaboration.md
Comment (johnwatson484) [reply guides/collaboration.md:11]: We don't use it on the Platform team, but it's usage is very common in the other teams.

File: standards/quick-reference.md
Comment (johnwatson484) [reply standards/quick-reference.md]: I've deliberately not mentioned the `Secret` resource in Kubernetes because it's the subject of a bit of debate on how secure it is (in Defra).

But the point I was trying to make here is, if the use of Pod Identity would allow you to avoid putting a secret into the cluster then you need to do it.  eg, we should have database passwords in our cluster.  I've reworded it to make it clearer.

File: standards/source-code.md
Comment (johnwatson484) [reply standards/source-code.md:57]: Good idea, I've added a `resources` folder and added it there.

File: standards/version-control.md
Comment (johnwatson484) [reply standards/version-control.md:17]: What would be the benefit of keeping lock file in sync?  Is it good practice or would it potentially cause an issue if we don't?

**Comment by johnwatson484** [comment]
> A lot of good content in here 👏. I like the definition of guides and standards. I think there is scope to have a level above a standard, something like a mandate or mandated standard i.e. if you want stuff to work it needs to meet the mandated standard. Some examples would be a branch name is a standard but not mandated, nothing isn't going to work that doesn't adhere to it. At least atm, that is how branches work, ofc, in the future it might be the case that only branches meeting that standard will be built and that would be the time when it becomes mandates. Another example would be the naming of the repos, that is really a mandated standard as things will not work if it isn't followed. Something to discuss outside of GitHub issues most likely.
> 
> I've added comments about typos and broken links where I've spotted them. There is also some rogue whitespace I noticed when I looked at the changes locally but I've not commented as it isn't visible in the GitHub diff.
> 
> A slight aside, I wonder if there should be a style guide for markdown content as part of this repo. My preference for content is the same as code i.e. a max of 80 character wide lines. The majority of content I see doesn't appear to have any limitation. Obviously there are pros and cons to both so whatever is decided I'm OK with, it is more about having a standard. This goes for things like blank lines between headings, double spaces for line breaks, capitalisation of DEFRA, bullets not ending with a period, etc.
> What do you think? Something else for wider discussion perhaps?

In terms of the standards, I see all of them as mandated.  They're all traced back to decisions we've made as a programme over the last two years.  Not adhering to some may not cause failure, but following them provides other benefits we want to realise in terms of developer mobility and agility.  For example, the branch naming convention we agreed as it means that developers already know the way a team will work if they change/support other teams and it removes the need for teams to debate over a naming policy if we've got a central one for the programme.  If we learn through our teams that convention is too restrictive or doesn't work for them, then we can discuss it as a programme.  If we agree it shouldn't be mandated then I'd rather move it to guidance and have it as a recommendation of what we suggest you do.

I've mirrored the structure of standards and guidance on the Defra one where standards are you must do this, guidance is you could do as well as examples of how.

The naming of GitHub repos is partially mandated outside the programme, ie our Principal developers have said we must prefix with `ffc` so it's easier to trace ownership.  That works fine for us as it means we can apply that pattern to our secret scanning.  It also allowed us to build on it to include the workstream as a sub prefix (is that a term?) so we can easily identify ownership of a repository within the programme or if we need more specific "stuff" applied to a workstream we are more likely going to be able to do it.  Nothing would break (other than secret scanning) if we didn't follow this standard, but we want to make sure teams do.  The naming also came from what we agreed with our Cloud architect (when we had one on the team) with regards naming of Azure repos for when we needed private repositories.  So far we haven't.

On style guide, yes I totally agree we should have something.  I'll create a ticket for the backlog to update the documentation section in this repo with one.  As the content in this repo has been written by many people with no agreed style guide, it was very inconsistent.  In this PR I've standardised so it's as consistent as possible based on the format of the majority.  But that's not to say it's best practice nor will it stay consistent if we don't have something.

On the capitilisation of DEFRA, don't :-p.  Government style guide states that it should be written in title case, ie "Defra".  Below is taken from their guide you can see at https://www.gov.uk/guidance/style-guide/a-to-z-of-gov-uk-style

## Capitals for government departments
Use the following conventions for government departments. A department using an ampersand in its logo image is fine but use ‘and’ when writing in full text.
...
Department for Environment, Food and Rural Affairs (Defra)

File: guides/collaboration.md
Comment (paulsimonandrews) [reply guides/collaboration.md:11]: True. It is used heavily on Grants.

File: standards/backlog-management.md
Comment (paulsimonandrews) [inline standards/backlog-management.md]: I don't think the development guide standards should apply to/impact the wider Agile teams. For instance, on Grants they don't use Jira, but have chosen as a team to use Trello for their Alpha. 

I agree that teams of developers within a wider agile team should use Jira to track **development tickets**, but this shouldn't be mandated for the whole agile team.

Maybe be explicit about which "team" is being referred to.

File: standards/code-review.md
Comment (paulsimonandrews) [inline standards/code-review.md]: Might be worth stating that the scope of the PR will (should) be stated in the acceptance criteria of a JIRA ticket?

File: standards/version-control.md
Comment (st3v3nhunt) [reply standards/version-control.md:17]: A cursory Google doesn't find anything interesting.
IME there are no problems with it being out of sync and is quite common for that to be the case. It is mostly the case for our repos and it hasn't caused any issues.
My preference is to have them in sync and I tend to do this but that is about the extent of doing it.

File: standards/developer-workflow.md
Comment (paulsimonandrews) [inline standards/developer-workflow.md]: I don't think TDD is appropriate for everything and shouldn't be mandated as a standard in all circumstances. For me TDD makes sense for developing production quality code, but not for quick throw-away prototypes to understand a problem/system which is very useful in the early stages of agile development (discovery, alpha).

I'd prefer to see something along the lines of "developer follows appropriate development strategy"  (this may be TDD where appropriate)

File: standards/documentation.md
Comment (paulsimonandrews) [inline standards/documentation.md]: "will mostly be that of an" -> "will mostly be within an"

**Comment by st3v3nhunt** [comment]
Good to know about the capitalisation. Now if only the GitHub org could be re-cased!

File: standards/jenkins.md
Comment (paulsimonandrews) [inline standards/jenkins.md:14]: I think this need clarifying. I think I know what you mean here, but incoming devs may be confused by this. Are we talking env vars set on Jenkins itself and microservice parameters set in Azure? Or parameters that get set to a Jenkins env var?

File: standards/kubernetes/helm-charts.md
Comment (paulsimonandrews) [reply standards/kubernetes/helm-charts.md:2]: Isn't `Kubenetes` the Geordie pronunciation?

File: standards/kubernetes/helm-charts.md
Comment (paulsimonandrews) [inline standards/kubernetes/helm-charts.md:13]: Might be worth saying this is set in `package.json` or linking to where this is described

File: standards/kubernetes/labels.md
Comment (paulsimonandrews) [inline standards/kubernetes/labels.md:28]: Should this be: "should be rejected" (there is no automatic rejection)

File: standards/kubernetes/priority.md
Comment (paulsimonandrews) [inline standards/kubernetes/priority.md:7]: Do we describe how this is specified anywhere?

File: standards/kubernetes/probes.md
Comment (paulsimonandrews) [inline standards/kubernetes/probes.md:4]: Drop-in default probes are provided by the FFC Helm chart library

File: standards/secrets-management.md
Comment (paulsimonandrews) [inline standards/secrets-management.md:4]: Is it worth saying server-side secret detection is in place too. Issues are raised in slack notifications and you are obliged to check this and assess any alerts on repos that you work on ASAP?

File: standards/secrets-management.md
Comment (paulsimonandrews) [inline standards/secrets-management.md:1]: This title doesn't match the content i.e. only secret detection is mentioned here. Should it be "Secret Detection"?

File: standards/source-code.md
Comment (paulsimonandrews) [inline standards/source-code.md:17]: This has changed I think. FFC team is added as `Write` and individuals can be added as `Admin` where appropriate.

File: standards/source-code.md
Comment (paulsimonandrews) [inline standards/source-code.md:28]: Remove the "as"?

Also could add branches are deleted automatically on PR merge

File: standards/version-control.md
Comment (johnwatson484) [reply standards/version-control.md:17]: Maybe we should leave that to the consuming teams to decide rather than enforcing it by default?

**Comment by johnwatson484** [comment]
> Good to know about the capitalisation. Now if only the GitHub org could be re-cased!

Ha! yes, it hurts my eyes every time I see it.  A lot of the style guide is about accessibility and what is perceived as easier to read and digest.  This is sometimes at odds with what is grammatically correct.

File: standards/testing.md
Comment (paulsimonandrews) [inline standards/testing.md:14]: Add this and next sentence as bullet point

File: standards/testing.md
Comment (paulsimonandrews) [inline standards/testing.md:79]: remove comma

File: standards/testing.md
Comment (paulsimonandrews) [inline standards/testing.md:67]: remove comma

File: standards/version-control.md
Comment (paulsimonandrews) [inline standards/version-control.md:3]: Might be worth saying somewhere that using the Jenkins pipeline library handles the Docker and Helm microservice versioning

File: guides/collaboration.md
Comment (paulsimonandrews) [inline guides/collaboration.md]: Rendering of lines 58 and 59 seems to run into each other. Not sure why

File: guides/collaboration.md
Comment (paulsimonandrews) [inline guides/collaboration.md]: Anchoralerts and secretdetection render on the same line too

File: standards/source-code.md
Comment (paulsimonandrews) [inline standards/source-code.md]: It's at `v0.14.3` now

File: guides/kubernetes/probes.md
Comment (paulsimonandrews) [inline guides/kubernetes/probes.md:7]: Probes are available in the helper templates of the FFC Helm chart library

**Review by paulsimonandrews — COMMENTED** [review]
This is looking great and turning into a very useful resource. Have made a bunch of suggestions none of which should block merging, but it would be good if you can consider the,

**Review by st3v3nhunt — APPROVED** [review]
LGTM 👍

**Pull request merged**
